### ISSUE TITLE
Principia DLL won't load

### ISSUE THREAD

**Issue by rsparkyc** [mockingbirdnest/Principia#4130]
![image](https://github.com/user-attachments/assets/7a18f315-6b23-4876-85f3-2ce2b2dca405)

Additional Context:
Running MacOS Sequoia, followed everything that was listed in discord beginning here:
https://discord.com/channels/319857228905447436/480397772248580098/1129098065563893781

Also started a conversation here about it:
https://discord.com/channels/319857228905447436/480397772248580098/1301195449968885760

note: the SO file never shows up as getting blocked in system settings, and yet I can see the file exists when I look at the path

**Comment by rsparkyc** [comment]
[KSP.log](https://github.com/user-attachments/files/17620536/KSP.log)

**Comment by pleroy** [comment]
Is there anything like `ktrace` or `truss` on macOS?  A trace of the system calls that happen when it's trying to launch the game might help.

**Comment by rsparkyc** [comment]
So I wrote this script:
```
#!/bin/bash

# Define the output file
OUTPUT_FILE="ksp_fs_usage.log"

# Start KSP
open -a /Applications/KSP_osx/KSP.app
echo "Launching KSP..."

# Start fs_usage and filter for "KSP", saving the output to a file
echo "Monitoring file and network activity with fs_usage and saving to $OUTPUT_FILE..."
sudo fs_usage | grep "KSP" > "$OUTPUT_FILE"
```
I've gzipped the output and attached it here: [ksp_fs_usage.log.gz](https://github.com/user-attachments/files/17624065/ksp_fs_usage.log.gz)

**Comment by rsparkyc** [comment]
Interestingly, I saw these lines in that file:
```
318469:16:31:27  stat64            /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so                   0.000008   KSP
318470:16:31:27  access            /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so                   0.000002   KSP
```

I tried dropping the so file in there, and it seems to have loaded

**Comment by rsparkyc** [comment]
![image](https://github.com/user-attachments/assets/2dbcddce-9b06-4850-89db-f5bded62ba55)

Seems to work! The question is, why did I need to put in in there?

**Comment by rnlahaye** [comment]
Oddly, I do see `stat64            /Applications/KSP_osx/GameData/Principia/MacOS64/principia.so` in the log also.

**Comment by rsparkyc** [comment]
Yeah, and the error specifically stated that was what it couldn't find

**Comment by rnlahaye** [comment]
Silly question: does the file have executable permissions? (What does `ls -lh /Applications/KSP_osx/GameData/Principia/MacOS64/principia.so` say?)

**Comment by rsparkyc** [comment]
it does not:
```
-rw-rw-r--@ 1 ryancaskey  staff    25M Oct 29 19:11 ./principia.so
```

after a `chmod +x principia.so` we get this:
```
-rwxrwxr-x@ 1 ryancaskey  staff    25M Oct 29 19:11 ./principia.so
```

however, the problem persists (unless, of course, I drop the unadulterated file into `KSP_osx/KSP.app/Contents/Frameworks`)

**Comment by astro-angelfish** [comment]
I've run into similar issue either. Seems the loader didn't cooperate well along with my OS:

Here's my log: https://pastebin.com/FYiXGyBY

**Comment by pleroy** [comment]
That last comment is some kind of Linux, so unrelated to the rest of the issue.  I updated the title to clarify that this issue is about macOS.

**Comment by astro-angelfish** [comment]
> That last comment is some kind of Linux (we are not told which distro), so unrelated to the rest of the issue. I updated the title to clarify that this issue is about macOS.

Ok. I'm gonna open another new issue for it with more detailed information.

**Comment by BaconCat1** [comment]
I am also having this issue, however dropping the file in /Applications/KSP_osx/GameData/Principia/MacOS64/principia.so does not work. Adding it here also does not help either : /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so 
I also ran the shell script to obtain the logs, and that path never even appeared. 

[ksp_fs_usage.log.zip](https://github.com/user-attachments/files/22062788/ksp_fs_usage.log.zip)

Edit: And yes, I am on macOS Sequoia, not some random linux distro nobody cares about.

**Comment by rsparkyc** [comment]
> I also ran the shell script to obtain the logs, and that path never even appeared.

I actually do see that in your log file:
```
16:31:27  stat64            /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so                   0.000008   KSP         
16:31:27  access            /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so                   0.000002   KSP         
```

However, I tried firing this up with the latest version (‎Lamé), and am seeing the same behavior now as well.

**Comment by rsparkyc** [comment]
after some more chatGPT prompting I got it to work:
https://chatgpt.com/share/68cc0dac-e37c-8001-984b-8769d2fd816b

Had to link the so files with an x86_64 build of libc

**Comment by pleroy** [comment]
That ChatGPT thread is mightily long and has a lot of alternatives to explore.  Is there a short TL;DR of the things to do to fix the problem?

**Comment by rsparkyc** [comment]
It is, but I intentionally left it all because I didn't know what step actually did the trick. But what I _think_ did it was the following:

```
/usr/bin/arch -x86_64 /bin/bash -lc '/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"'

/usr/bin/arch -x86_64 /bin/bash -lc 'brew install llvm@17 || brew install llvm'

chmod u+w "/Applications/KSP_osx/GameData/Principia/MacOS64/principia.so"
chmod u+w "/Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/c++/libc++.1.dylib \
  /usr/lib/libc++.1.dylib \
  "/Applications/KSP_osx/GameData/Principia/MacOS64/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/c++/libc++abi.1.dylib \
  /usr/lib/libc++abi.dylib \
  "/Applications/KSP_osx/GameData/Principia/MacOS64/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/unwind/libunwind.1.dylib \
  /usr/local/opt/llvm@17/lib/libunwind.1.dylib \
  "/Applications/KSP_osx/GameData/Principia/MacOS64/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/c++/libc++.1.dylib \
  /usr/lib/libc++.1.dylib \
  "/Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/c++/libc++abi.1.dylib \
  /usr/lib/libc++abi.dylib \
  "/Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so"

install_name_tool -change /usr/local/opt/llvm/lib/unwind/libunwind.1.dylib \
  /usr/local/opt/llvm@17/lib/libunwind.1.dylib \
  "/Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so"
```

Honestly, it may only require doing those steps to the `principia.so` file that I've placed within KSP.app. It would be good to get some others to try to see if they can boil this down to the minimum set of steps.

Edit: forgot `/usr/bin/arch -x86_64 /bin/bash -lc 'brew install llvm@17 || brew install llvm'`

**Comment by eggrobin** [comment]
It would be interesting to know whether it still works if you get rid of the copy of principia.so in /Applications/KSP_osx/KSP.app/Contents/Frameworks/principia.so.

That one is probably not going to be found in Lánczos (to be released Sunday), because of #4308.

**Comment by pleroy** [comment]
I think that after the first step (which installs `brew` for `x86_64`) there has to be an extra step of installing `llvm@17` for `x86_64`.  We have to do something like that for building Principia on our M1, see [the macOS instructions](https://github.com/mockingbirdnest/Principia/blob/master/documentation/Setup.md#macos).

I take it that you are running on an ARM machine?

**Comment by rsparkyc** [comment]
I am using an arm mac (m2) 

I think i forgot this command:
```
/usr/bin/arch -x86_64 /bin/bash -lc 'brew install llvm@17 || brew install llvm'
```

going to update my post

**Comment by pleroy** [comment]
I have added words to the [FAQs](https://github.com/mockingbirdnest/Principia/wiki/Installing,-reporting-bugs,-and-frequently-asked-questions#mac-users), focusing on the case where the shared library is in the canonical location.  (The reason why you had to have a copy in `KSP.app/Contents/Frameworks` has never been clear to me, and I would rather not invite users to do the copy unless they have to; this is going to be a sure mess when they upgrade, especially with upcoming changes to the way we load the shared library.)

**Comment by adigyran** [comment]
'/Users/adigyran/Library/Application Support/Steam/steamapps/common/Kerbal Space Program/KSP.app'

my ksp in here so those fixes not work at all :( moved so file, installed llvm and such and still this error. m3 max sequoia

**Comment by adigyran** [comment]
<img width="947" height="651" alt="Image" src="https://github.com/user-attachments/assets/a97bf022-6c21-4842-b15e-071612941f64" />

**Comment by adigyran** [comment]
installing llvm 20 helped. Please update faq

**Comment by alcarithemad** [comment]
TL;DR: this is the minimal set of commands you need to run to fix the files. As a player, you don't need `brew` or the matching `llvm` release.
(edit: to be explicit, I did this, it worked, I have the Lehmer release loaded in KSP right now)
```
cd GameData/Principia/macOS/
install_name_tool -change /usr/local/opt/llvm@20/lib/c++/libc++abi.1.dylib /usr/lib/libc++abi.dylib x64/principia.so
install_name_tool -change /usr/local/opt/llvm@20/lib/unwind/libunwind.1.dylib /usr/lib/system/libunwind.dylib x64/principia.so
install_name_tool -change /usr/local/opt/llvm@20/lib/c++/libc++.1.dylib /usr/lib/libc++.1.dylib x64/principia.so
install_name_tool -change /usr/local/opt/llvm@20/lib/c++/libc++abi.1.dylib /usr/lib/libc++abi.dylib x64_AVX_FMA/principia.so
install_name_tool -change /usr/local/opt/llvm@20/lib/unwind/libunwind.1.dylib /usr/lib/system/libunwind.dylib x64_AVX_FMA/principia.so
install_name_tool -change /usr/local/opt/llvm@20/lib/c++/libc++.1.dylib /usr/lib/libc++.1.dylib x64_AVX_FMA/principia.so
```

I'd suggest doing this at build/packaging time but I'm not familiar with how your current macOS build system works. Probably you'd invoke the drop-in equivalent `llvm-install-name-tool` to munge the paths as above or perhaps add a sprinkling of `-Wl,-rpath,<path>` arguments matching the above targets to your `CXXFLAGS`.

Long version:

macOS shared libraries contain paths (often absolute paths) to their dynamically linked dependencies. the release artifacts have `rpath`s for the C++ runtime libraries that point explicitly to a homebrew-provided LLVM install. macOS ships all 3 of the libraries in question, but several release cycles ago the actual files for most system libraries were absorbed into a giant blob called the DYLD cache and no longer appear in their notional file system locations, so what path you need to ask for is... not obvious. Some of the earlier attempts in this issue almost got this right, but ended up still needing llvm's `libunwind` because it's in a different subdirectory for no good reason.

In case another one becomes required some day, or someone searching the web needs to know:

You can check what these paths currently are with a command like `dyld_info -linked_dylibs x64*/principia.so`.

The macOS SDK contains metadata files you can use to identify the notional paths these live at:
```
> cd /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/
> grep 'install-name.*libc++\.' -r .
./libc++.1.tbd:install-name:    '/usr/lib/libc++.1.dylib'
> grep 'install-name.*libunwind' -r .
./libSystem.B.tbd:install-name:    '/usr/lib/system/libunwind.dylib'
./system/libunwind.tbd:install-name:    '/usr/lib/system/libunwind.dylib'
./libSystem.B_asan.tbd:install-name:    '/usr/lib/system/libunwind.dylib'
> grep 'install-name.*libc++abi' -r .
./libc++abi.tbd:install-name:    '/usr/lib/libc++abi.dylib'
```

**Comment by pleroy** [comment]
If it works for you, fine, but this is certainly not a solution that can be generally recommended.

The root of the problem is *not* what paths the shared library references.  The root of the problem is the version of the C++ libraries that are present on your system.  Apple has been notorious for lagging behind LLVM by months if not *years*.  The latest version of LLVM is 22.1.4.  Xcode went from 19.1.5 to 21.1.6 on March 18, 2026, with the release of Xcode 26.4.  Prior to that date, they didn't have support for our binaries, which are built with LLVM 20.

If you have XCode 26.4 or later on your system, what you did is fine.  If you don't, then things may break in mysterious ways and I don't want to debug them.

**Comment by alcarithemad** [comment]
If that's the case then the instructions on the FAQ page seem incoherent, as they direct one to use Apple's `libc++` and `libc++abi` with LLVM 20's `libunwind`? If it's worth installing LLVM, why not point all three at it?

**Comment by pleroy** [comment]
I believe that you are right but then I don't have access to a mac, so I am hesitant to edit instructions that I cannot test.  The current FAQs come from [this comment above](https://github.com/mockingbirdnest/Principia/issues/4130#issuecomment-3308450664), I think.  But then the FAQs are a wiki, so anyone can edit them.
